iceoryx_utils/source/concurrent/dual_access_transaction_tray.cpp

```diff
@@ -0,0 +1,95 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_utils/internal/concurrent/dual_access_transaction_tray.hpp"
+
+namespace iox
+{
+namespace concurrent
+{
+DualAccessTransactionTray::AccessGuard::AccessGuard(DualAccessTransactionTray& transactionTray, AccessToken accessToken)
+    : m_transactionTray(transactionTray)
+    , m_accessToken(accessToken)
+{
+    transactionTray.acquireExclusiveAccess(m_accessToken);
+}
+
+DualAccessTransactionTray::AccessGuard::~AccessGuard()
+{
+    m_transactionTray.releaseExclusiveAccess(m_accessToken);
+}
+
+void DualAccessTransactionTray::cleanupAndSyncMemory(const AccessToken tokenToCleanup)
+{
+    releaseExclusiveAccess(tokenToCleanup);
+    // just to be sure the memory is synchronized
+    AccessToken expected = tokenToCleanup;
+    m_accessToken.compare_exchange_strong(expected, AccessToken::NONE, std::memory_order_acq_rel);
+}
+
+void DualAccessTransactionTray::acquireExclusiveAccess(AccessToken tokenToAcquireAccess)
+{
+    auto existingToken = m_accessToken.exchange(tokenToAcquireAccess, std::memory_order_acquire);
+    if (existingToken == tokenToAcquireAccess)
+    {
+        // TODO return expected DOUBLE_ACQUIRE_BROKEN_INVARIANT
+        std::terminate();
+    }
+    else if (existingToken != AccessToken::NONE)
+    {
+        if (tokenToAcquireAccess == AccessToken::LEFT)
+        {
+            if (m_waitingLineLeft.wait().has_error())
+            {
+                // TODO return expected ERROR_WHILE_WAITING_FOR_SEMAPHORE
+            }
+        }
+        else
+        {
+            if (m_waitingLineRight.wait().has_error())
+            {
+                // TODO return expected ERROR_WHILE_WAITING_FOR_SEMAPHORE
+            }
+        }
+    }
+}
+
+void DualAccessTransactionTray::releaseExclusiveAccess(AccessToken tokenToBeReleased)
+{
+    AccessToken expected = tokenToBeReleased;
+    auto casSuccessful = m_accessToken.compare_exchange_strong(expected, AccessToken::NONE, std::memory_order_release);
+    if (!casSuccessful)
+    {
+        if (expected == AccessToken::NONE)
+        {
+            // TODO return expected RELEASE_ON_NONE_BROKEN_INVARIANT
+            std::terminate();
+        }
+        else if (tokenToBeReleased == AccessToken::LEFT)
+        {
+            // post can result in either EINVAL or EOVERFLOW; neither of those can be triggered by this code
+            IOX_DISCARD_RESULT(m_waitingLineRight.post());
```

Comment:
We post even though no one might be waiting? This is dangerous. Consider a Sequence of acq LEFT, rel LEFT, acq RIGHT, rel RIGHT without someone waiting. Afterwards a LEFT and RIGHT request could arrive concurrently and both would gain access (both semaphore counters are 1) or some other problematic state I think. 

The compare exchanges may protect against this, I need to fully think this through, but the potential interleaves are non-trivial. In any case, there might be a problem but I would like to find a convincing argument why there is none (or a fix if there is).

Reply:
With the current implementation two are needed else we experience kind of an ABA problem. Let's assume we have only one semaphore, LEFT has the lock and RIGHT is waiting on the semaphore. When LEFT does a sem_post and immediately tries to get the lock, the `exchange` will tell LEFT that the lock is held by RIGHT and must therefore do a sem_wait. It can happen that RIGHT was not yet scheduled and the semaphore count is still 1, which means that LEFT will continue its execution and won't do a sem_post at the end, leaving RIGHT in a deadlock.
With an alternative implementation this problem might be solved differently.